File: Raytracing/raytracercsharp/raytracer/Scene.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace raytracer
{
    class Scene
    {
        public List<Renderable> renderables = new List<Renderable>();
        public List<Light> lights = new List<Light>();
        public Vector3 ambientLight;
        public Vector3 skyColour = new Vector3(0.1f, 0.3f, 0.8f);
        public Camera camera = new Camera();
        public bool shadows = true;
        public int recurseLimit = 4;
        public SFML.Graphics.Image skyTexture;

        public HitPoint intersect(Ray ray)
        {
            HitPoint hp = new HitPoint();
            foreach (Renderable r in renderables)
            {
                hp.nearest(r.intersect(ray));
            }
            return hp;
        }

        public Vector3 trace(float x, float y)
        {
            Ray ray = camera.viewToRay(x, y);
            return trace(ray, 0);
        }

        float remap(float src_range_start, float src_range_end, float dst_range_start, float dst_range_end, float value_to_remap)
        {
            return ((dst_range_end - dst_range_start) * (value_to_remap - src_range_start)) / (src_range_end - src_range_start) + dst_range_start;
        }

        public Vector3 trace(Ray ray, int recurseDepth)
        {
            Vector3 output = skyColour;
            if (recurseDepth > recurseLimit)
            {
                output.set(0, 0, 0);
                return output;
            }

            if (skyTexture != null)
            {
                // Get the colour of the skybox that this ray would hit.
                int tx = (int)(remap(-3.14159265f, 3.14159265f, 0, skyTexture.Size.X - 1.0f, (float)Math.Atan2(ray.delta.x, ray.delta.z)));
                int ty = (int)(remap(1.0f, -1.0f, 0, skyTexture.Size.Y - 1.0f, ray.delta.y));
                SFML.Graphics.Color c = skyTexture.GetPixel((uint)tx, (uint)ty);
                output.x = c.R / 255.0f;
                output.y = c.G / 255.0f;
                output.z = c.B / 255.0f;
            }

            HitPoint hp = intersect(ray);
            if (hp.hit)
            {
                Vector3 specColour = new Vector3(0, 0, 0);
                Vector3 diffuseColour = new Vector3(0, 0, 0);

                // Loop through every light in the scene.
                foreach (Light light in lights)
                {
                    // Direction and distance from the intersect point to the current light.
                    Vector3 lightDirection = light.position - hp.position;
                    float lightDistance = lightDirection.normalise();

                    Vector3 lightColour = light.colour;
                    if (shadows)
                    {
                        // Cast a ray from the intersect point to each light. If anything is in the way, a shadow is being cast on the intersect point.
                        Ray shadowRay = new Ray();
                        shadowRay.start = hp.position + lightDirection * 0.001f; // The lightDirection*0.001f bit is to stop floating point error making the ray collide with it's starting point.
                        shadowRay.delta = lightDirection;
                        HitPoint shadowhp = intersect(shadowRay);

                        // If a shadow is detected, the light's contribution is zero.
                        if (shadowhp.hit && shadowhp.distance < lightDistance)
                        {
                            lightColour.set(0, 0, 0);
                        }
                    }

                    // Calculate standard lambert diffuse lighting
                    float lambert = hp.normal.dot(lightDirection);
                    if (lambert < 0)
                        lambert = 0;

                    // Accumulate diffuse lighting.
                    diffuseColour = diffuseColour + hp.renderable.material.diffuse * lightColour * lambert;

                    // Accumulate specular lighting.
                    Vector3 rl = lightDirection.reflect(hp.normal);
                    float specular = rl.dot(ray.delta);
                    if (specular > 0)
                    {
                        specColour += hp.renderable.material.specular * light.colour * (float)Math.Pow(specular, hp.renderable.material.specularPower);
                    }
                }

                Vector3 reflectionColour = new Vector3();
                Ray reflectionRay = new Ray();
                // If the hit object's material has reflection, cast a new ray recursively.
                if (hp.renderable.material.reflective.length() > 0)
                {
                    reflectionRay.delta = (ray.delta.reflect(hp.normal)).normalised();
                    reflectionRay.start = hp.position + reflectionRay.delta * 0.001f;
                    reflectionColour = trace(reflectionRay, recurseDepth + 1) * hp.renderable.material.reflective;
                }

                // Final output is diffuse + specular + reflection.
                output = ambientLight + diffuseColour + specColour + reflectionColour;
            }
            return output;
        }
    }
}

File: Raytracing/raytracer [Slight optimisation]/raytracer/scene.cpp
#include "scene.h"
#include "kf/kf_algorithms.h"

Scene::Scene() :m_ambient(0.0f, 0.0f, 0.0f), m_sky(0.1f, 0.3f, 0.8f), m_shadows(true), m_recurseLimit(4)
{
//	m_skytexture.load_bmp("KKZ5V.bmp");
}

Scene::~Scene()
{
	for (unsigned int i = 0; i < m_lights.size(); ++i)
	{
		delete m_lights[i];
	}
	for (unsigned int i = 0; i < m_renderables.size(); ++i)
	{
		delete m_renderables[i];
	}
}

HitPoint Scene::intersect(const kf::Ray &ray)
{
	HitPoint hp;

	for (unsigned int i = 0; i < m_renderables.size(); ++i)
	{
		if (m_renderables[i]->m_active)
		{
			// Find the nearest intersect point.
			hp.nearest(m_renderables[i]->intersect(ray));
		}
	}
	return hp;
}

kf::Colour Scene::trace(float x, float y)
{
	kf::Colour output;
	kf::Ray ray = m_camera.viewToRay(x, y);
	output = trace(ray, 0);

	return output;
}

kf::Colour Scene::trace(const kf::Ray &ray, int recurseDepth)
{
	kf::Colour output = m_sky;

	if (recurseDepth > m_recurseLimit)
	{
		output.set(0, 0, 0);
		return output;
	}

	if (!m_skytexture.is_empty())
	{
		// Get the colour of the skybox that this ray would hit.
		int tx = int(kf::remap<float>(-3.14159265f, 3.14159265f, 0, m_skytexture.width() - 1.0f, atan2(ray.delta().x, ray.delta().z)));
		int ty = int(kf::remap<float>(1.0f, -1.0f, 0, m_skytexture.height() - 1.0f, ray.delta().y));
		output.r = *m_skytexture.data(tx, ty, 0, 0) / 255.0f;
		output.g = *m_skytexture.data(tx, ty, 0, 1) / 255.0f;
		output.b = *m_skytexture.data(tx, ty, 0, 2) / 255.0f;
		output.toLinear();
	}

	// Find the intersect with the scene.
	HitPoint hp = intersect(ray);
	// If true, we hit something. Otherwise return the skybox.
	if (hp.m_hit)
	{
		kf::Colour specColour(0, 0, 0);
		kf::Colour diffuseColour(0, 0, 0);

		// Loop through every light in the scene.
		for (unsigned int i = 0; i < m_lights.size(); ++i)
		{
			// Direction and distance from the intersect point to the current light.
			kf::Vector3 lightDirection = m_lights[i]->m_position - hp.m_position;
			float lightDistance = lightDirection.normalise();

			kf::Colour lightColour = m_lights[i]->m_colour;
			if (m_shadows)
			{
				// Cast a ray from the intersect point to each light. If anything is in the way, a shadow is being cast on the intersect point.
				kf::Ray shadowRay;
				shadowRay.start(hp.m_position + lightDirection*0.001f); // The lightDirection*0.001f bit is to stop floating point error making the ray collide with it's starting point.
				shadowRay.delta(lightDirection);
				HitPoint shadowhp = intersect(shadowRay);

				// If a shadow is detected, the light's contribution is zero.
				if (shadowhp.m_hit && shadowhp.m_distance < lightDistance)
				{
					lightColour.set(0, 0, 0);
				}
			}

			// Calculate standard lambert diffuse lighting
			float lambert = hp.m_normal.dot(lightDirection);
			if (lambert < 0)
				lambert = 0;

			// Accumulate diffuse lighting.
			diffuseColour = diffuseColour + hp.m_renderable->m_material.m_diffuse * lightColour * lambert;

			// Accumulate specular lighting.
			kf::Vector3 rl = lightDirection.reflect(hp.m_normal);
			float specular = rl.dot(ray.delta());
			if (specular > 0)
			{
				specColour += hp.m_renderable->m_material.m_specular * m_lights[i]->m_colour * pow(specular, hp.m_renderable->m_material.m_specularPower);
			}
		}

		kf::Colour reflectionColour(0,0,0);
		kf::Ray reflectionRay;
		// If the hit object's material has reflection, cast a new ray recursively.
		if (hp.m_renderable->m_material.m_reflective.length() > 0)
		{
			reflectionRay.delta(normalise(ray.delta().reflect(hp.m_normal)));
			reflectionRay.start(hp.m_position + reflectionRay.delta()*0.001f);
			reflectionColour = trace(reflectionRay, recurseDepth + 1) * hp.m_renderable->m_material.m_reflective;
		}

		// Final output is diffuse + specular + reflection.
		output = m_ambient*1 + diffuseColour*1 + specColour*1 + reflectionColour*1;
	}

	return output;
}

